crates/shared/src/sources/balancer_v2/pools/common.rs

```diff
@@ -59,9 +67,9 @@ impl<Factory> PoolInfoFetcher<Factory> {
     }
 
     /// Returns a Balancer base pool contract instance at the specified address.
-    fn base_pool_at(&self, pool_address: H160) -> BalancerV2BasePool {
+    fn base_pool_at(&self, pool_address: &H160) -> Arc<BalancerV2BasePool> {
         let web3 = self.vault.raw_instance().web3();
-        BalancerV2BasePool::at(&web3, pool_address)
+        crate::get_or_init!(BalancerV2BasePool, POOLS, pool_address, &web3)
```

Comment:
I find this quite ugly (it adds global state + a somewhat hard to understand macro call). How much time is this saving? Would it be sufficient to keep POOLS cache on self or does self get re-created? In any case maybe this should move into a separate PR?

Reply:
> How much time is this saving?

I captured a flamegraph before and after the optimization which indicates that it makes sense to have it as the `CONTRACT::at()` function calls are not cheap because we have to hash all the function signatures and events emitted by the contract for every single liquidity fetch request.
I'll check if we can move the cached contracts into `Self`. But I think we can't use a generic function for this so we either have to duplicate the function a bunch of times or live with the macro.

![flamegraph_before](https://github.com/cowprotocol/services/assets/19190235/b1d915f6-b277-47dd-af8b-29cda6ff30f0)
![flamegraph_after](https://github.com/cowprotocol/services/assets/19190235/e560f72e-3565-4b29-95e0-1dfd257a0111)

Reply:
From the flam graph it's not obvious to me what the savings are. Would you mind highlighting it?

Reply:
We spend less time in `CONTRACT::with_deployment_info()` which is the function called by `CONTRACT::at()` (pretty high on the big stack in the middle, you can see it by hovering over segments when opening the file in a separate tab).
But since this caching in particular is a point of contention in this PR and could be done in a separate PR I will drop it for now so we can merge the over all more important aspects of this PR.